### PR TITLE
Implement Backpropagation learning

### LINKED ISSUE #22: Implement learning
- [x] Prepare Backpropagation which will use Gradient Descent
- [x] Implement Gradient Descent

### PR DESCRIPTION
**Pull request by gustav1101**
On popular demand: Here is the current state of the backpropagation. What is still in parts missing is the evaluation of test data and all debugging.

There will probably be some bugs in the math-stuff as well, I was really tired when I wrote some of this. Also needs to update with master branch.

Closes #22

### CODE REVIEW

**Comment by gustav1101** [comment]
This should actually be merged now. There network is not learning correctly yet, but since all others need this code and it throws no errors or warnings I would suggest to check and then merge it.
